### PR TITLE
Expose remote attestation endpoint for webclients

### PR DESCRIPTION
**Pull request by jul-sh**
This adds an endpoint for future web clients to communicate with the runtime directly, by exposing a [grpc-web](https://github.com/grpc/grpc-web) endpoint. grpc-web is an adaption of the gRPC protocol that supports wire formats that web clients can use. 

We'll be using the binary format, see the OSS description here: https://github.com/grpc/grpc-web#wire-format-mode. Defacto this means that requests are sent as unary HTTP/1 requests, with the body consisting of binary encoded protobuf messages. 

Imo this is a nice approach to supporting OSS web clients in the linux-binary version of the runtime. The runtime itself will expose compatible endpoints, with no need to spin up a separate translation layer server. 

In prod deployments too secret to discuss in depth here we'd use a translation service. 
In the bare-metal version of the runtime the companion app will expose this endpoint.

### CODE REVIEW

File: oak_functions/loader/src/grpc.rs [hunk 1]
```diff
@@ -82,21 +82,34 @@ pub async fn create_and_start_grpc_server<F: Future<Output = ()>>(
...
+            )
+            .context("Couldn't create remote attestation server")?,
+        );
+
+    // Create a server and add the relevant service defintions. Server creation
+    // and start is handled entirely witin each respective conditional arm, as
+    // the added service alters the type signature of the created server.
+    if cfg!(feature = "oak-web") {
+        tonic::transport::Server::builder()
+            .add_service(grpc_unary_attestation_service.clone())
```
Comment (conradgrobler) [inline oak_functions/loader/src/grpc.rs:102]: Is it required to add the service twice? I would have expected `enable` to add support for grpc-web requests without breaking normal gRPC.

File: oak_functions/loader/src/grpc.rs [hunk 2]
```diff
@@ -82,21 +82,34 @@ pub async fn create_and_start_grpc_server<F: Future<Output = ()>>(
...
+                additional_info,
+                ErrorLogger { logger },
+            )
+            .context("Couldn't create remote attestation server")?,
+        );
+
+    // Create a server and add the relevant service defintions. Server creation
+    // and start is handled entirely witin each respective conditional arm, as
+    // the added service alters the type signature of the created server.
+    if cfg!(feature = "oak-web") {
```
Comment (conradgrobler) [inline oak_functions/loader/src/grpc.rs:100]: I think you need to use `#[cfg(....)]`. The `tonic_web` reference would not be available if the oak-web feature is not enabled.

File: oak_functions/loader/src/grpc.rs [hunk 3]
```diff
@@ -82,21 +82,34 @@ pub async fn create_and_start_grpc_server<F: Future<Output = ()>>(
...
+        tonic::transport::Server::builder()
```
Comment (conradgrobler) [inline oak_functions/loader/src/grpc.rs:101]: I think this needs `.accept_http1(true)` as there is no TLS, so the browser won't use ALPN to negotiate the HTTP version.

File: oak_functions/loader/src/grpc.rs [hunk 4]
```diff
@@ -27,6 +27,7 @@ use oak_logger::OakLogger;
 use oak_utils::LogError;
 use prost::Message;
 use std::{future::Future, net::SocketAddr};
+use tonic_web;
```
Comment (conradgrobler) [inline oak_functions/loader/src/grpc.rs:30]: I don't think this is needed.

**Review by conradgrobler — APPROVED** [review]

File: oak_functions/loader/src/grpc.rs (hunk 3 above)
Comment (jul-sh) [reply oak_functions/loader/src/grpc.rs:101]: good point, overlooked that indeed. added, thanks.

File: oak_functions/loader/src/grpc.rs (hunk 1 above)
Comment (jul-sh) [reply oak_functions/loader/src/grpc.rs:103]: You're right, the documentation on it isn't super clear, but it does pass through non grpc-web requests to the underlying service. 

https://github.com/hyperium/tonic/blob/350ccf95e4ee71f85a2a1af7dc3e1df6cca1a2e3/tonic-web/src/service.rs#L158

I probably would've preferred if the enable method didn't do that (or exposed an option to not do that), but it is what it is. Updated.

**Comment by jul-sh** [comment]
@conradgrobler thanks for the review. :)

File: oak_functions/loader/src/grpc.rs [hunk 5]
```diff
@@ -82,21 +82,35 @@ pub async fn create_and_start_grpc_server<F: Future<Output = ()>>(
...
+                additional_info,
+                ErrorLogger { logger },
+            )
+            .context("Couldn't create remote attestation server")?,
+        );
+
+    // Create a server and add the relevant service defintions. Server creation
+    // and start is handled entirely witin each respective conditional arm, as
+    // the added service alters the type signature of the created server.
+    if cfg!(feature = "oak-web") {
```
Comment (conradgrobler) [inline oak_functions/loader/src/grpc.rs:99]: nit: you can remove the `if` statement if you mark the statement in the else clause with `#[cfg(not(...))]`.

**Pull request merged**